### PR TITLE
feat: Add context parallel support for Qwen3.5 MoE

### LINKED ISSUE #1498: Add context parallel support for Qwen3.5
**Is your feature request related to a problem? Please describe.**
Qwen 3.5 uses Gated DeltaNet which is not currently supported for cp

**Describe the solution you'd like**
Requires to have cp support so that to run RL training with longer sequence. https://github.com/NVIDIA-NeMo/RL/issues/2025

### PR DESCRIPTION
**Pull request by zpqiu**
# What does this PR do ?

Add CP support for Qwen3.5 MoE.

# Changelog

- Add specific line by line info of high level changes in this PR.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)
#1498 

## Convergence Experiment

- Pink: CP1
- Orange: CP2

截屏2026-03-13 12 49 40

```yaml
# Qwen3.5 MoE 35B LLM SFT — CP=2 experiment
# Usage:
#   torchrun --nproc-per-node=8 examples/llm_finetune/finetune.py \
#     --config examples/llm_finetune/qwen/qwen3_5_moe_35b_cp2.yaml

step_scheduler:
    global_batch_size: 128
    local_batch_size: 4
    ckpt_every_steps: 500
    val_every_steps: 10
    num_epochs: 1
    max_steps: 200

dist_env:
    backend: nccl
    timeout_minutes: 60

rng:
    _target_: nemo_automodel.components.training.rng.StatefulRNG
    seed: 1234
    ranked: true

model:
    _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_pretrained
    pretrained_model_name_or_path: Qwen/Qwen3.5-35B-A3B-Base
    backend:
        _target_: nemo_automodel.components.models.common.BackendConfig
        attn: te
        linear: te
        rms_norm: torch_fp32
        experts: te
        dispatcher: no
        rope_fusion: false
        fake_balanced_gate: false
        enable_hf_state_dict_adapter: true

checkpoint:
    enabled: false

distributed:
    strategy: fsdp2
    tp_size: 1
    cp_size: 2
    pp_size: 1
    ep_size: 8

    sequence_parallel: false
    activation_checkpointing: true

loss_fn:
    _target_: nemo_automodel.components.loss.masked_ce.MaskedCrossEntropy

dataset:
    _target_: nemo_automodel.components.datasets.llm.chat_dataset.ChatDataset
    path_or_dataset_id: allenai/tulu-3-sft-mixture
    split: train
    shuffle_seed: 1234
    seq_length: 2048
    padding: max_length
    truncation: true

packed_sequence:
    packed_sequence_size: 0

dataloader:
    _target_: torchdata.stateful_dataloader.StatefulDataLoader
    collate_fn: nemo_automodel.components.datasets.utils.default_collater
    shuffle: true

validation_dataset:
    _target_: nemo_automodel.components.datasets.llm.chat_dataset.ChatDataset
    path_or_dataset_id: allenai/tulu-3-sft-mixture
    split: "train[:1024]"
    shuffle_seed: 1234
    seq_length: 2048
    padding: max_length
    truncation: true

validation_dataloader:
    _target_: torchdata.stateful_dataloader.StatefulDataLoader
    collate_fn: nemo_automodel.components.datasets.utils.default_collater
    batch_size: 4

optimizer:
    _target_: torch.optim.Adam
    betas: [0.9, 0.999]
    eps: 1e-7
    lr: 1.0e-5
    weight_decay: 0
    foreach: false
```

### CODE REVIEW

**Comment by zpqiu** [comment]
The two CI failures are unrelated to this PR:
                                                                                                                                                                                                                                                           
  1. test_lora / test_module_matcher — Regression introduced by 20a91bfa ("improve ux of peft") on main. This PR does not touch any peft/lora/module_matcher files.
  2. Check imports (NGC CUDA) — import torch fails due to a CUDA library version mismatch in the CI container (libcusparse.so.12: undefined symbol: __nvJitLinkGetErrorLogSize_12_9). This is a CI environment issue.

File: tests/functional_tests/context_parallel/L2_CP_Qwen3_5MoE_LinearAttn_Test.sh [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -xeuo pipefail # Exit immediately if a command exits with a non-zero status
+
+export PYTHONPATH=${PYTHONPATH:-}:$(pwd)
+export CUDA_VISIBLE_DEVICES="0,1"
+
+# Run Qwen3.5 MoE linear attention CP test with 2 GPUs
+torchrun --nproc_per_node=2 --nnodes=1 \
```
Comment (akoumpa) [inline tests/functional_tests/context_parallel/L2_CP_Qwen3_5MoE_LinearAttn_Test.sh]: I guess we need coverage here and an easy way to test loss against cp=1 🤔

Comment (zpqiu) [reply tests/functional_tests/context_parallel/L2_CP_Qwen3_5MoE_LinearAttn_Test.sh]: I referred to this CP test, `tests/functional_tests/context_parallel/run_attention_cp.py`, to add a loss alignment check at here: https://github.com/NVIDIA-NeMo/Automodel/pull/1560/changes/BASE..8e7059343e947e632d5c63c62d0f862025518489#diff-11157b9a9e96003f138a571acb63b53096a8832eb30ad2b1fffc6cff5845867cR232-R246

Comment (zpqiu) [reply tests/functional_tests/context_parallel/L2_CP_Qwen3_5MoE_LinearAttn_Test.sh]: Do you have any other suggestions for coverage tests?

Comment (akoumpa) [reply tests/functional_tests/context_parallel/L2_CP_Qwen3_5MoE_LinearAttn_Test.sh]: Hi @zpqiu , 

coverage is easy to fix:
```
TRANSFORMERS_OFFLINE=1 \
python \
-m torch.distributed.run --nproc_per_node=2 --nnodes=1 \ # adjust as necessary
-m coverage run \
-m pytest tests/functional_tests/checkpoint/test_hf_consolidated_vlm.py \ # change this to your test
```

Thanks for the pointer, on the cp loss test 🙇

**Review by akoumpa — APPROVED** [review]

**Comment by akoumpa** [comment]
Thanks a lot @zpqiu for the contribution, very much appreciated. Please feel free to LMK if there's anything I can help with.

**Comment by HKAB** [comment]
Hey @zpqiu, I'm using v0.4.0 and maybe the `nemo_automodel/components/models/qwen3_next/model.py` file didn't get merge yet so training with cp_size > 1 will throw error in https://github.com/NVIDIA-NeMo/Automodel/blob/v0.4.0/nemo_automodel/components/models/qwen3_5_moe/cp_linear_attn.py#L369.
